File: src/main/java/cn/academy/vanilla/teleporter/skills/ShiftTeleport.java
/**
* Copyright (c) Lambda Innovation, 2013-2016
* This file is part of the AcademyCraft mod.
* https://github.com/LambdaInnovation/AcademyCraft
* Licensed under GPLv3, see project root for more information.
*/
package cn.academy.vanilla.teleporter.skills;

import cn.academy.ability.api.Skill;
import cn.academy.ability.api.ctrl.ActionManager;
import cn.academy.ability.api.ctrl.SkillInstance;
import cn.academy.ability.api.ctrl.action.SkillSyncAction;
import cn.academy.ability.api.data.AbilityData;
import cn.academy.vanilla.teleporter.client.TPParticleFactory;
import cn.academy.vanilla.teleporter.entity.EntityMarker;
import cn.academy.vanilla.teleporter.util.TPSkillHelper;
import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.util.generic.MathUtils;
import cn.lambdalib.util.generic.RandUtils;
import cn.lambdalib.util.generic.VecUtils;
import cn.lambdalib.util.helper.Color;
import cn.lambdalib.util.helper.Motion3D;
import cn.lambdalib.util.mc.EntitySelectors;
import cn.lambdalib.util.mc.Raytrace;
import cn.lambdalib.util.mc.WorldUtils;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.block.Block;
import net.minecraft.command.IEntitySelector;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.Vec3;
import net.minecraftforge.common.util.ForgeDirection;

import java.util.ArrayList;
import java.util.List;

import static cn.lambdalib.util.generic.MathUtils.*;

/**
 * @author WeAthFolD
 */
@Registrant
public class ShiftTeleport extends Skill {

    static final Color CRL_BLOCK_MARKER = new Color().setColor4i(139, 139, 139, 180),
            CRL_ENTITY_MARKER = new Color().setColor4i(235, 81, 81, 180);

    public static final ShiftTeleport instance = new ShiftTeleport();

    private ShiftTeleport() {
        super("shift_tp", 4);
    }

    static float getExpIncr(int attackEntities) {
        return (1 + attackEntities) * 0.002f;
    }

    private static float getDamage(float exp) {
        return lerpf(16, 27, exp);
    }

    private static float getRange(float exp) {
        return lerpf(25, 35, exp);
    }

    private static float getConsumption(float exp) {
        return lerpf(374, 500, exp);
    }

    private static float getOverload(float exp) {
        return lerpf(38, 27, exp);
    }

    @Override
    public SkillInstance createSkillInstance(EntityPlayer player) {
        return new SkillInstance().addChild(new ShiftTPAction())
                .setEstmCP(getConsumption(AbilityData.get(player).getSkillExp(instance)));
    }

    public static class ShiftTPAction extends SkillSyncAction {

        float exp;

        public ShiftTPAction() {
            super(-1);
        }

        @Override
        public void onStart() {
            super.onStart();

            exp = aData.getSkillExp(instance);

            ItemStack stack = player.getCurrentEquippedItem();
            Block block;
            if (!(stack != null && stack.getItem() instanceof ItemBlock
                    && (Block.getBlockFromItem(stack.getItem())) != null))
                ActionManager.abortAction(this);

            if (isRemote)
                startEffects();
        }

        @Override
        public void onTick() {
            if (isRemote)
                updateEffects();
        }

        boolean attacked;

        @Override
        public void onEnd() {
            if (isRemote)
                return;

            Block block;
            ItemStack stack = player.getCurrentEquippedItem();
            attacked = stack != null && stack.getItem() instanceof ItemBlock
                    && (block = Block.getBlockFromItem(stack.getItem())) != null;
            if (!attacked)
                return;

            ItemBlock item = (ItemBlock) stack.getItem();
            MovingObjectPosition position = getTracePosition();

            if (item.field_150939_a.canPlaceBlockAt(player.worldObj, position.blockX, position.blockY, position.blockZ)
                    && cpData.perform(getOverload(exp), getConsumption(exp))) {

                item.placeBlockAt(stack, player, player.worldObj, position.blockX, position.blockY, position.blockZ,
                        position.sideHit, (float) position.hitVec.xCoord, (float) position.hitVec.yCoord,
                        (float) position.hitVec.zCoord, stack.getItemDamage());

                if (!player.capabilities.isCreativeMode) {
                    if (--stack.stackSize <= 0)
                        player.setCurrentItemOrArmor(0, null);
                }

                List<Entity> list = getTargetsInLine();
                for (Entity target : list) {
                    TPSkillHelper.attack(player, instance, target, getDamage(exp));
                }

                player.worldObj.playSoundAtEntity(player, "academy:tp.tp_shift", 0.5f, 1f);
                aData.addSkillExp(instance, getExpIncr(list.size()));

                if (!player.capabilities.isCreativeMode) {
                    if (stack.stackSize-- == 0) {
                        player.setCurrentItemOrArmor(0, null);
                    }
                }

                setCooldown(instance, (int) lerpf(20, 5, exp));
            }
        }

        @Override
        public void onFinalize() {
            if (isRemote)
                endEffects();
        }

        // TODO: Some boilerplate... Clean this up in case you aren't busy
        private int[] getTraceDest() {
            double range = getRange(exp);
            MovingObjectPosition result = Raytrace.traceLiving(player, range, EntitySelectors.nothing);
            if (result != null) {
                ForgeDirection dir = ForgeDirection.values()[result.sideHit];
                return new int[] { result.blockX + dir.offsetX, result.blockY + dir.offsetY,
                        result.blockZ + dir.offsetZ };
            }
            Motion3D mo = new Motion3D(player, true).move(range);
            return new int[] { (int) mo.px, (int) mo.py, (int) mo.pz };
        }

        private MovingObjectPosition getTracePosition() {
            double range = getRange(exp);
            MovingObjectPosition result = Raytrace.traceLiving(player, range, EntitySelectors.nothing);
            if (result != null) {
                ForgeDirection dir = ForgeDirection.values()[result.sideHit];
                result.blockX += dir.offsetX;
                result.blockY += dir.offsetY;
                result.blockZ += dir.offsetZ;
                return result;
            }
            Motion3D mo = new Motion3D(player, true).move(range);
            return new MovingObjectPosition((int) mo.px, (int) mo.py, (int) mo.pz, 0,
                    VecUtils.vec(mo.px, mo.py, mo.pz));
        }

        private List<Entity> getTargetsInLine() {
            int[] dest = getTraceDest();
            Vec3 v0 = VecUtils.vec(player.posX, player.posY, player.posZ),
                    v1 = VecUtils.vec(dest[0] + .5, dest[1] + .5, dest[2] + .5);

            AxisAlignedBB area = WorldUtils.minimumBounds(v0, v1);
            IEntitySelector selector = new IEntitySelector() {

                @Override
                public boolean isEntityApplicable(Entity entity) {
                    double hw = entity.width / 2;
                    return VecUtils.checkLineBox(VecUtils.vec(entity.posX - hw, entity.posY, entity.posZ - hw),
                            VecUtils.vec(entity.posX + hw, entity.posY + entity.height, entity.posZ + hw), v0,
                            v1) != null;
                }

            };
            return WorldUtils.getEntities(player.worldObj, area,
                    EntitySelectors.and(EntitySelectors.living, EntitySelectors.excludeOf(player), selector));
        }

        // CLIENT
        @SideOnly(Side.CLIENT)
        EntityMarker blockMarker;

        @SideOnly(Side.CLIENT)
        List<EntityMarker> targetMarkers;

        @SideOnly(Side.CLIENT)
        int effTicker;

        @SideOnly(Side.CLIENT)
        void startEffects() {
            targetMarkers = new ArrayList();
            if (isLocal()) {
                blockMarker = new EntityMarker(player.worldObj);
                blockMarker.ignoreDepth = true;
                blockMarker.width = blockMarker.height = 1.2f;
                blockMarker.color = CRL_BLOCK_MARKER;
                blockMarker.setPosition(player.posX, player.posY, player.posZ);

                player.worldObj.spawnEntityInWorld(blockMarker);
            }
        }

        @SideOnly(Side.CLIENT)
        void updateEffects() {
            if (isLocal()) {
                if (++effTicker == 3) {
                    effTicker = 0;
                    for (EntityMarker em : targetMarkers) {
                        em.setDead();
                    }
                    targetMarkers.clear();
                    List<Entity> targetsInLine = getTargetsInLine();
                    for (Entity e : targetsInLine) {
                        EntityMarker em = new EntityMarker(e);
                        em.color = CRL_ENTITY_MARKER;
                        em.ignoreDepth = true;
                        player.worldObj.spawnEntityInWorld(em);
                        targetMarkers.add(em);
                    }
                }

                int[] dest = getTraceDest();
                blockMarker.setPosition(dest[0] + 0.5, dest[1], dest[2] + 0.5);
            }
        }

        @SideOnly(Side.CLIENT)
        void endEffects() {
            if (isLocal()) {
                for (EntityMarker em : targetMarkers)
                    em.setDead();
                blockMarker.setDead();
            }

            if (attacked) {
                int[] dest = getTraceDest();
                double dx = dest[0] + .5 - player.posX, dy = dest[1] + .5 - (player.posY - 0.5),
                        dz = dest[2] + .5 - player.posZ;
                double dist = MathUtils.length(dx, dy, dz);
                Motion3D mo = new Motion3D(player.posX, player.posY - 0.5, player.posZ, dx, dy, dz);
                mo.normalize();

                double move = 1;
                for (double x = move; x <= dist; x += (move = RandUtils.ranged(0.6, 1))) {
                    mo.move(move);
                    player.worldObj.spawnEntityInWorld(TPParticleFactory.instance.next(player.worldObj, mo.getPosVec(),
                            VecUtils.vec(RandUtils.ranged(-.05, .05), RandUtils.ranged(-.02, .05),
                                    RandUtils.ranged(-.05, .05))));
                }
            }
        }

    }
}

File: src/main/scala/cn/academy/vanilla/teleporter/skill/ShiftTeleport.scala
/**
  * Copyright (c) Lambda Innovation, 2013-2016
  * This file is part of the AcademyCraft mod.
  * https://github.com/LambdaInnovation/AcademyCraft
  * Licensed under GPLv3, see project root for more information.
  */
package cn.academy.vanilla.teleporter.skill

import java.util.function.Predicate

import cn.academy.ability.api.{AbilityContext, Skill}
import cn.academy.ability.api.context.{ClientContext, ClientRuntime, Context, RegClientContext}
import cn.academy.ability.api.data.AbilityData
import cn.academy.vanilla.teleporter.client.TPParticleFactory
import cn.academy.vanilla.teleporter.entity.EntityMarker
import cn.academy.vanilla.teleporter.util.TPSkillHelper
import cn.lambdalib.annoreg.core.Registrant
import cn.lambdalib.s11n.network.NetworkMessage.Listener
import cn.lambdalib.util.generic.{MathUtils, RandUtils, VecUtils}
import cn.lambdalib.util.helper.{Color, Motion3D}
import cn.lambdalib.util.mc.{EntitySelectors, Raytrace, WorldUtils}
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.block.Block
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.{ItemBlock, ItemStack}
import net.minecraft.util.{AxisAlignedBB, MovingObjectPosition, Vec3}
import net.minecraftforge.common.util.ForgeDirection

/**
  * @author WeAthFolD, KSkun
  */
object ShiftTeleport extends Skill("shift_tp", 4) {

  expCustomized = true

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyID: Int): Unit = activateSingleKey(rt, keyID, p => new STContext(p))

  override def getSkillExp(data: AbilityData): Float = {
    if(AbilityContext.of(data.getEntity, this).isEntirelyDisableBreakBlock) 1f else data.getSkillExp(this)
  }

}

object STContext {

  final val MSG_EXECUTE = "execute"

}

import cn.lambdalib.util.generic.MathUtils._
import cn.academy.ability.api.AbilityAPIExt._
import STContext._

class STContext(p: EntityPlayer) extends Context(p, ShiftTeleport) {

  private val exp = ctx.getSkillExp

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.SERVER))
  private def s_madeAlive() = {
    val stack: ItemStack = player.getCurrentEquippedItem
    val block: Block = null
    if(!(stack != null && stack.getItem.isInstanceOf[ItemBlock] && Block.getBlockFromItem(stack.getItem) != null))
      terminate()
  }

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  private def l_onKeyUp() = {
    sendToServer(MSG_EXECUTE)
  }

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  private def l_onKeyAbort() = {
    sendToSelf(MSG_EXECUTE, false.asInstanceOf[AnyRef])
    terminate()
  }

  private var attacked: Boolean = false

  @Listener(channel=MSG_EXECUTE, side=Array(Side.SERVER))
  private def s_execute(): Unit = {
    val stack: ItemStack = player.getCurrentEquippedItem
    val block: Block = Block.getBlockFromItem(stack.getItem)
    attacked = stack != null && stack.getItem.isInstanceOf[ItemBlock] && block != null
    if(!attacked) return

    sendToClient(MSG_EXECUTE, attacked.asInstanceOf[AnyRef])

    val item: ItemBlock = stack.getItem.asInstanceOf[ItemBlock]
    val position: MovingObjectPosition = getTracePosition

    if(item.field_150939_a.canPlaceBlockAt(player.worldObj, position.blockX, position.blockY, position.blockZ) && ctx.canBreakBlock(player.worldObj, position.blockX, position.blockY, position.blockZ) && ctx.consume(getOverload(exp), getConsumption(exp))) {
      item.placeBlockAt(stack, player, player.worldObj, position.blockX, position.blockY, position.blockZ, position.sideHit, position.hitVec.xCoord.toFloat, position.hitVec.yCoord.toFloat, position.hitVec.zCoord.toFloat, stack.getItemDamage)
      if(!player.capabilities.isCreativeMode) if( {
        stack.stackSize -= 1; stack.stackSize
      } <= 0) player.setCurrentItemOrArmor(0, null)
      val list: java.util.List[Entity] = getTargetsInLine
      import scala.collection.JavaConversions._
      for(target <- list) {
        TPSkillHelper.attack(ctx, target, getDamage(exp))
      }
      player.worldObj.playSoundAtEntity(player, "academy:tp.tp_shift", 0.5f, 1f)
      ctx.addSkillExp(getExpIncr(list.size))
      ctx.setCooldown(lerpf(100, 60, exp).toInt)
    }
    terminate()
  }

  private def getExpIncr(attackEntities: Int): Float = (1 + attackEntities) * 0.002f

  private def getDamage(exp: Float): Float = lerpf(15, 35, exp)

  private def getRange(exp: Float): Float = lerpf(25, 35, exp)

  private def getConsumption(exp: Float): Float = lerpf(260, 320, exp)

  private def getOverload(exp: Float): Float = lerpf(40, 30, exp)

  // TODO: Some boilerplate... Clean this up in case you aren't busy
  def getTraceDest: Array[Int] = {
    val range: Double = getRange(exp)
    val result: MovingObjectPosition = Raytrace.traceLiving(player, range, EntitySelectors.nothing)
    if(result != null) {
      val dir: ForgeDirection = ForgeDirection.values().apply(result.sideHit)
      return Array[Int](result.blockX + dir.offsetX, result.blockY + dir.offsetY, result.blockZ + dir.offsetZ)
    }
    val mo: Motion3D = new Motion3D(player, true).move(range)
    Array[Int](mo.px.toInt, mo.py.toInt, mo.pz.toInt)
  }

  def getTracePosition: MovingObjectPosition = {
    val range: Double = getRange(exp)
    val result: MovingObjectPosition = Raytrace.traceLiving(player, range, EntitySelectors.nothing)
    if(result != null) {
      val dir: ForgeDirection = ForgeDirection.values().apply(result.sideHit)
      result.blockX += dir.offsetX
      result.blockY += dir.offsetY
      result.blockZ += dir.offsetZ
      return result
    }
    val mo: Motion3D = new Motion3D(player, true).move(range)
    new MovingObjectPosition(mo.px.toInt, mo.py.toInt, mo.pz.toInt, 0, VecUtils.vec(mo.px, mo.py, mo.pz))
  }

  def getTargetsInLine: java.util.List[Entity] = {
    val dest: Array[Int] = getTraceDest
    val v0: Vec3 = VecUtils.vec(player.posX, player.posY, player.posZ)
    val v1: Vec3 = VecUtils.vec(dest(0) + .5, dest(1) + .5, dest(2) + .5)
    val area: AxisAlignedBB = WorldUtils.minimumBounds(v0, v1)
    val pred: Predicate[Entity] = EntitySelectors.living.and(EntitySelectors.exclude(player)).and(new Predicate[Entity] {

      override def test(entity: Entity): Boolean = {
        val hw = entity.width / 2
        VecUtils.checkLineBox(VecUtils.vec(entity.posX - hw, entity.posY, entity.posZ - hw),
          VecUtils.vec(entity.posX + hw, entity.posY + entity.height, entity.posZ + hw), v0, v1) != null
      }

    })
    WorldUtils.getEntities(player.worldObj, area, pred)
  }

}

@Registrant
@SideOnly(Side.CLIENT)
@RegClientContext(classOf[STContext])
class STContextC(par: STContext) extends ClientContext(par) {

  private val CRL_BLOCK_MARKER: Color = new Color().setColor4i(139, 139, 139, 180)
  private val CRL_ENTITY_MARKER: Color = new Color().setColor4i(235, 81, 81, 180)

  private var blockMarker: EntityMarker = _
  private var targetMarkers: java.util.List[EntityMarker] = _
  private var effTicker: Int = 0

  @Listener(channel=MSG_MADEALIVE, side=Array(Side.CLIENT))
  private def l_start() = {
    targetMarkers = new java.util.ArrayList[EntityMarker]
    if(isLocal) {
      blockMarker = new EntityMarker(player.worldObj)
      blockMarker.ignoreDepth = true
      blockMarker.height = 1.2f
      blockMarker.width = 1.2f
      blockMarker.color = CRL_BLOCK_MARKER
      blockMarker.setPosition(player.posX, player.posY, player.posZ)
      player.worldObj.spawnEntityInWorld(blockMarker)
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def l_tick() = {
    if(isLocal) {
      if({effTicker += 1; effTicker} == 3) {
        effTicker = 0
        import scala.collection.JavaConversions._
        for(em <- targetMarkers) {
          em.setDead()
        }
        targetMarkers.clear()
        val targetsInLine: java.util.List[Entity] = par.getTargetsInLine
        import scala.collection.JavaConversions._
        for(e <- targetsInLine) {
          val em: EntityMarker = new EntityMarker(e)
          em.color = CRL_ENTITY_MARKER
          em.ignoreDepth = true
          player.worldObj.spawnEntityInWorld(em)
          targetMarkers.add(em)
        }
      }
      val dest: Array[Int] = par.getTraceDest
      blockMarker.setPosition(dest(0) + 0.5, dest(1), dest(2) + 0.5)
    }
  }

  @Listener(channel=MSG_TERMINATED, side=Array(Side.CLIENT))
  private def l_terminated() = {
    if(isLocal) {
      import scala.collection.JavaConversions._
      for(em <- targetMarkers) em.setDead()
      if(blockMarker != null) blockMarker.setDead()
    }
  }

  @Listener(channel=MSG_EXECUTE, side=Array(Side.CLIENT))
  private def c_end(attacked: Boolean) = {
    if(isLocal) {
      import scala.collection.JavaConversions._
      for(em <- targetMarkers) em.setDead()
      blockMarker.setDead()
    }

    if(attacked) {
      val dest: Array[Int] = par.getTraceDest
      val dx: Double = dest(0) + .5 - player.posX
      val dy: Double = dest(1) + .5 - (player.posY - 0.5)
      val dz: Double = dest(2) + .5 - player.posZ
      val dist: Double = MathUtils.length(dx, dy, dz)
      val mo: Motion3D = new Motion3D(player.posX, player.posY - 0.5, player.posZ, dx, dy, dz)
      mo.normalize
      var move: Double = 1
      var x: Double = move
      while(x <= dist) {
        {
          mo.move(move)
          player.worldObj.spawnEntityInWorld(TPParticleFactory.instance.next(player.worldObj, mo.getPosVec, VecUtils.vec(RandUtils.ranged(-.05, .05), RandUtils.ranged(-.02, .05), RandUtils.ranged(-.05, .05))))
        }
        move = RandUtils.ranged(0.6, 1)
        x += move
      }
    }
  }

}
